src/client-side-metrics/metrics-collector.ts

```diff
@@ -0,0 +1,329 @@
+  private receivedFirstResponse: boolean;
+  private metricsHandlers: IMetricsHandler[];
+  private firstResponseLatency: number | null;
+  private serverTimeRead: boolean;
+  private serverTime: number | null;
+  private dateProvider: DateProvider;
+
+  /**
+   * @param {ITabularApiSurface} tabularApiSurface Information about the Bigtable table being accessed.
+   * @param {IMetricsHandler[]} metricsHandlers The metrics handlers used for recording metrics.
+   * @param {string} methodName The name of the method being traced.
+   * @param {string} projectId The id of the project.
+   * @param {DateProvider} dateProvider A provider for date/time information (for testing).
+   */
+  constructor(
+    tabularApiSurface: ITabularApiSurface,
+    metricsHandlers: IMetricsHandler[],
+    methodName: string,
+    projectId?: string,
+    dateProvider?: DateProvider
+  ) {
+    this.zone = undefined;
+    this.cluster = undefined;
+    this.tabularApiSurface = tabularApiSurface;
+    this.methodName = methodName;
+    this.operationStartTime = null;
+    this.attemptStartTime = null;
+    this.receivedFirstResponse = false;
+    this.metricsHandlers = metricsHandlers;
+    this.firstResponseLatency = null;
+    this.serverTimeRead = false;
+    this.serverTime = null;
+    this.projectId = projectId;
+    if (dateProvider) {
+      this.dateProvider = dateProvider;
+    } else {
+      this.dateProvider = new DefaultDateProvider();
+    }
+  }
+
+  /**
+   * Assembles the basic attributes for metrics. These attributes provide
+   * context about the Bigtable environment and the operation being performed.
+   * @param {string} projectId The Google Cloud project ID.
+   * @returns {Attributes} An object containing the basic attributes.
+   */
+  private getBasicAttributes(projectId: string) {
+    return {
+      projectId,
+      instanceId: this.tabularApiSurface.instance.id,
+      table: this.tabularApiSurface.id,
+      cluster: this.cluster,
+      zone: this.zone,
+      appProfileId: this.tabularApiSurface.bigtable.appProfileId,
+      methodName: this.methodName,
+      clientName: `nodejs-bigtable/${version}`,
+    };
+  }
+
+  /**
+   * Assembles the attributes for operation latency metrics.  These attributes
+   * provide context about the Bigtable environment, the operation being performed, and the final status of the operation.
+   * Includes whether the operation was a streaming operation or not.
+   * @param {string} projectId The Google Cloud project ID.
+   * @param {OperationOnlyAttributes} operationOnlyAttributes The attributes of the operation.
+   * @returns {OnOperationCompleteAttributes} An object containing the attributes for operation latency metrics.
+   */
+  private getOperationLatencyAttributes(
+    projectId: string,
+    operationOnlyAttributes: OperationOnlyAttributes
+  ): OnOperationCompleteAttributes {
+    return Object.assign(
+      operationOnlyAttributes,
+      this.getBasicAttributes(projectId)
+    );
+  }
```

Comment:
Can you explain this a bit more? I'm having a hard time following all the different layers (StandardAttributes/OperationOnlyAttributes/onOperationCompleteAttributes/getOperationLatencyAttributes/getBasicAttributes)

Especially since they are spread across multiple files. It seems a bit much

Reply:
With the new metrics handler API the call stack can be simplified so I removed getBasicAttributes, getOperationAttributes, getAttemptAttributes.